### PR TITLE
Thymesisflow software library and connections manager

### PR DESCRIPTION
**Pull request by mgazz**
libthymesisflow is the software library providing configuration capabilities for Thymesisflow.

The main features are:

1. Simple APIs to attach and detach remote memory
2. Agent keeping track of current connections
3. Command line interface to abstract low level implementation details
4. Simple interface to monitor latency and rx/tx bandwidth

More information provided in the README

### CODE REVIEW

File: libthymesisflow/README.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+```
+./bin/thymesisf-agent
+```
+
+To launch unit tests, run the following command.
+
+```
+./testbin/unittest
+```
```
Comment (christian-pinto) [inline libthymesisflow/README.md:51]: It is not clear that before starting the unit tests you have to start the `thymesisf-agent`. Please specify or change this behavior.

File: libthymesisflow/README.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+If not, you can find the instructions to [bring up the design](https://github.com/OpenCAPI/ThymesisFlow#how-to-bring-up-the-design).
+
+While the [mock](#build-mock) build is used for testing, in this scenario each requests will configure compute or memory node AFUs.
+For this scenario, recompile agent and CLI as following. 
+
+```
+make
+```
+
+**Note:** use `DEBUG=1` is required for running unittests successfully
```
Comment (christian-pinto) [inline libthymesisflow/README.md]: Is this true for the mock build too? It looks like it is not needed when building mock libraries.

File: libthymesisflow/README.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+* `MEMORY_SIZE` represents the amount of memory (in bytes) to be reserved. This needs to be a multiple of 256 MB
+* `EFFECTIVE_ADDRESS` address returned as `EA` during the [memory side setup](# Setup Memory side)
+
+#### Stats counters
+
+To keep track of the traffic on the channel, Thymesisflow provides three counters reporting:
+
+* `rx`: number of flits read from the channel
+* `rx`: number of flits transmitted on the channel
+* `latency`: average latency on the channel
```
Comment (christian-pinto) [inline libthymesisflow/README.md]: Is there a unit of measure for the latency metric? If yes, please specify.

File: libthymesisflow/cmd/thymesisf-agent.c [hunk 4]
```diff
@@ -0,0 +1,150 @@
+#include "proto.h"
```
Comment (christian-pinto) [inline libthymesisflow/cmd/thymesisf-agent.c]: There is no copyright notice in any of the source files, please add to all files missing it (both .h and .c).

File: libthymesisflow/src/thymesisflow.c [hunk 5]
```diff
@@ -0,0 +1,243 @@
+/// @file thymesisflow.c
+#include "thymesisflow.h"
+#include <fcntl.h>
+
+int detach(const uint64_t memory_size) {
```
Comment (christian-pinto) [inline libthymesisflow/src/thymesisflow.c]: I would consider renaming this method to something more descriptive. The name `detach` could be too generic for a code base dealing with attaching and detaching memory.

File: libthymesisflow/src/thymesisflow.c [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+                  iport_list *ports, const uint64_t size, uint64_t *eaptr) {
+
+    connection *conn = new_conn(circuit_id, afu_name, size);
+
+    add_conn(conn);
+
+#ifdef MOCK
+    log_info_ext("mocking memory connection by allocating only 256 MB...\n");
+
+    if (posix_memalign((void **)&conn->ea, 128, (256 >> 20)) != 0) {
```
Comment (christian-pinto) [inline libthymesisflow/src/thymesisflow.c]: Defining a macro for `128` could help other developers understanding what's that value about.

**Review by christian-pinto — CHANGES_REQUESTED** [review]

**Comment by mgazz** [comment]
Thank you @christian-pinto for the review. These are all valid points. I'll address them and update the pull requests.

**Review by christian-pinto — APPROVED** [review]

**Comment by christian-pinto** [comment]
All looks good to me. Just merged, thanks!

**Pull request merged**